rfcs/modal-component.md

```diff
@@ -0,0 +1,154 @@
+* `isOpen: boolean` - Required
+  Controls whether the modal is visible or not
+* `closable: boolean` - Optional, Defaults to `true`
+  Whether the modal should be closable by the user (either via escape, backdrop click, etc). You can set this to `false` if your modal has an action that the user _must_ take before closing.
+* `onClose: function` - Optional
+  A callback that is invoked when the modal will close
+* `size: SIZE.{small|medium|large|full|cover|auto}|string|number`, Defaults to SIZE.default
+  Controls the size of the modal (primarily width). Can be a SIZE constant or css `width` property value.
+* `aria-labelledby: string` - Optional, Default to ''
+  If specified, will be passed to the dialog element
+* `aria-describedby: string` - Optional, Defaults to ''
+  If specified, will be passed to the dialog element
+* `aria-label: string` - Optional, Defaults to ''
+  If specified, will be passed to the dialog element (described the modal for screen readers)
+* `id: string` - Optional, Defaults to ''
+  If specified, will be passed to the dialog element
+* `role: 'dialog'|'alertdialog'|string` - Optional, Defaults to 'dialog'
+  Which accessibility role this modal should have
+* `mountNode: HTMLElement` - Optional, Defaults to document.body
+  Where to mount the modal
+* `overrides: {Root, Backdrop, Dialog, DialogInner, Close}` - Optional
+  Overrides for presentational components. See "Presentational Components Props API" below.
+  * `[ComponentName]: ReactComponent | {props: {}, style: {}, component: ReactComponent}` - Optional
+
+### <ModalHeader/> API
+
+TBD
+
+### <ModalBody/> API
+
+TBD
+
+### <ModalFooter/> API
+
+TBD
+
+### SIZE Constant
+
+* `small` - 400px width
+* `medium` - 550px width
+* `large` - 800px width
+* `full` - Span full viewport, minus some margin
+* `cover` - Span full viewport, no margin
+* `auto` - Based on content width
+
+^ Open to ideas here. Too many options?
+
+### Presentational Components Props API
+
+Next properties are passed to every presentational (styled) component
+
+* `$isOpen: boolean`
+* `$size: string|number`
+* `$role: string`
+* `$closable: boolean`
+* `$theme: theme`
+
+More props likely TBA as development continues
+
+### Usage
+
+Basic usage:
+
+```javascript
+import * as React from 'react';
+import {Modal, SIZE} from 'baseui/modal';
+
+class App extends React.Component {
+  state = {
+    open: true
+  };
+  render() {
+    return (
+      <div>
+        <Button onClick={() => this.setState({open: true})}>Open</Button>
+        <Modal
+          isOpen={this.state.open}
+          close={() => this.setState({open: false})}
+        >
+          {({close}) => (
+            <>
+              <ModalTitle>Whatsup!</ModalTitle>
+              <ModalBody>This is a modal</ModalBody>
+              <ModalFooter>
```

Comment:
footer and header may be parts of body. Otherwise we may need to sync their styles (margin\padding) in most cases.

Reply:
There will be another wrapper element internally (StyledDialog or StyledDialogInner) that has the outer margin/padding to create spacing around the content. ModalTitle may have a marginBottom, and ModalFooter may have a marginTop, but otherwise these components won't have many styles.

Reply:
I think this looks good.
Some designers like to have a line separator that spans from one end to the other between the actual body and footer.
I assume I'd be able to "sneak" this line separator (`div`) between `ModalBody` and `ModalFooter` ?
```
<ModayBody>...</ModalBody>
<div style={{ ...some thin horizontal line styles...}} />
<ModalFooter>...</ModalFooter>
```